src/counts.jl

```diff
@@ -237,13 +237,62 @@ Add counts based on `x` to a count map. New entries will be added if new values
 If a weighting vector `wv` is specified, the sum of the weights is used rather than the
 raw counts.
 """
-function addcounts!(cm::Dict{T}, x::AbstractArray{T}) where T
+function addcounts!(cm::Dict{T}, x::AbstractArray{T}; uselessmem = false) where T
+    # if it's of bits type then can speed things up using radix sort
+    # from heuristics it was found that 2^16 = 65536 is the point at which
+    # radixsort is more performant
+    if uselessmem && radixsort_safe(T)
+        addcounts_radixsort!(cm, x)
+        return cm
+    else
+        addcounts_dict!(cm,x)
+        return cm
+    end
+end
+
+"""Dict-based addcounts method"""
+function addcounts_dict!(cm::Dict{T}, x::AbstractArray{T}) where T
     for v in x
-        cm[v] = get(cm, v, 0) + 1
+        index = Base.ht_keyindex2(cm, v)
+        if index > 0
+            @inbounds cm.vals[index] += 1
+        else
+            @inbounds Base._setindex!(cm, 1, v, -index)
+        end
+    end
+    return cm
+end
+
+"Can the type be sorted by radixsort"
+radixsort_safe(::Type{T}) where {T<:Number} = isbits(T)
```

Comment:
Even `<:Number` isn't safe, as custom `Number` types can be implemented by any package. For now, we should just hardcode the list of Base types which are known to be safe (in particular integers).

Also, the comment should really be "Can the type be sorted *safely* by radixsort".

I've filed https://github.com/JuliaCollections/SortingAlgorithms.jl/issues/26.